File: bst.c
#include <stdio.h>
#include <stdlib.h>

typedef struct _Node Node;
struct _Node {
	int info;
	Node *left;
	Node *right;
};

Node *insert(Node *p, int info) {
	if (p == NULL) {
		p = (Node *)malloc(sizeof(Node));
		p->info  = info;
		p->left  = NULL;
		p->right = NULL;
	} else {
		if (info < p->info) {
			p->left  = insert(p->left,  info);
		} else {
			p->right = insert(p->right, info);
		}
	}
	return p;
}

void inorder(Node *p) {
	if (p->left)  inorder(p->left);
	printf("%d\n", p->info);
	if (p->right) inorder(p->right);
}

int main() {
	Node *root = NULL;
	root = insert(root, 4);
	root = insert(root, 6);
	root = insert(root, 2);
	root = insert(root, 1);
	root = insert(root, 3);
	root = insert(root, 5);
	root = insert(root, 7);
	inorder(root);
	return 0;
}

File: bst.js
function Node(data) {
	this.data  = data;
	this.left  = null;
	this.right = null;
}

function insert(node, data) {
	if (node == null) {
		node = new Node(data);
		return node;
	} else {
		if (data < node.data) {
			node.left  = insert(node.left, data);
		} else {
			node.right = insert(node.right, data);
		}
		return node;
	}
}

function print(node) {
	if (node != null) {
		print(node.left);
		console.log(node.data);
		print(node.right);
	}
}

/*
var root = null;
for (var i = 0; i < 100; i++) {
	root = insert(root, parseInt(Math.random() * 100));
}
print(root);
*/
